### PR TITLE
LLVM 16 support

### LINKED ISSUE #38: Support for LLVM 16
Rust 1.70 uses LLVM 16 by default, so support here would be nice!

I'll try and do a PR similar to #25.

### PR DESCRIPTION
**Pull request by woodruffw**
WIP; closes #38.

### CODE REVIEW

File: src/types.rs [hunk 1]
```diff
@@ -153,6 +156,7 @@ impl Display for Type {
             Type::MetadataType => write!(f, "metadata"),
             Type::LabelType => write!(f, "label"),
             Type::TokenType => write!(f, "token"),
+            Type::TargetExtType => write!(f, "target"),
```
Comment (woodruffw) [inline src/types.rs:159]: This isn't quite right, since target types can take subtypes and integer values.

**Comment by woodruffw** [comment]
The main TODO here is support for target extension types: https://releases.llvm.org/16.0.0/docs/LangRef.html#target-extension-type

File: src/types.rs [hunk 2]
```diff
@@ -1080,6 +1164,11 @@ impl TypesBuilder {
             LLVMTypeKind::LLVMMetadataTypeKind => self.metadata_type(),
             LLVMTypeKind::LLVMLabelTypeKind => self.label_type(),
             LLVMTypeKind::LLVMTokenTypeKind => self.token_type(),
+            #[cfg(feature = "llvm-16-or-greater")]
+            LLVMTypeKind::LLVMTargetExtTypeKind => {
+                todo!()
```
Comment (woodruffw) [inline src/types.rs:1169]: From a quick look, there don't appear to be LLVM-C APIs for accessing this type's members, yet.

File: src/function.rs [hunk 3]
```diff
@@ -590,7 +590,9 @@ impl AttributesData {
             "builtin",
             "cold",
             "convergent",
+            #[cfg(feature = "llvm-15-or-lower")]
             "inaccessiblememonly",
+            #[cfg(feature = "llvm-15-or-lower")]
```
Comment (woodruffw) [inline src/function.rs:595]: I'm not sure about these (or the one below) -- I had to disable them on LLVM > 15 to get some of the tests passing, but I can't find any changesets indicating that these were actually removed.

**Comment by Benjins** [comment]
Hi there: thanks for all the work on the upgrade to LLVM 16! This PR has been a draft for a while, so I made a PR to your fork which should get the tests all passing (along with a merge from latest main). If you're okay with that, then the PR should (I think) be ready for review. Or if you want, I can pick up the upgrade work and make a new PR, if you have a lot else on your plate

Also, as you point out, Target Extension Types are not currently exposed to the C API, so until a future upgrade they won't be implementable. I don't know exactly how we should deal with them in the meantime, but my own opinion is that they shouldn't block this initial upgrade

Link to my PR for this fork:
https://github.com/trail-of-forks/llvm-ir/pull/1

**Comment by woodruffw** [comment]
Hi @Benjins, thanks! I'll take a look at your PR now; I greatly appreciate you picking it up.

**Comment by woodruffw** [comment]
Merged; I'll also get this updated against the upstream `main`. Thanks again @Benjins!

**Comment by woodruffw** [comment]
Tests for this are mostly passing for me now, although I'm getting a `SIGTRAP` on `llvm_10_tests` (this doesn't make any sense to me, but I haven't looked closely yet).

**Comment by Benjins** [comment]
> I'm getting a SIGTRAP on llvm_10_tests
If you're running tests with a debug rust build (i.e. not with the `--release` flag), there is an existing issue where LLVM's C API does not properly handle. I tend to test in release mode to avoid the crash. See https://github.com/cdisselkoen/llvm-ir/issues/44

**Comment by woodruffw** [comment]
Funny bug! Yeah, confirmed: `--release` fixes it, and all tests pass locally for me now.

**Comment by Benjins** [comment]
cc @cdisselkoen Not sure if you get a notification when a draft PR is un-drafted, but this PR should be ready for review. There are some follow-up changes that we'll probably want, and there is the question of what to do with Target Extension Types (not currently available via the C API). But this should provide initial LLVM 16 support that can be built on

**Review by cdisselkoen — APPROVED** [review]

**Comment by cdisselkoen** [comment]
Looks great. Thanks for everyone's work.

For follow-up, I think we can handle Target Extension Types much like we do things like BlockAddress today -- basically just have a marker that contains no data since we can't get any data from the C API.

**Pull request merged**